cilium/CFP-23380-ordered-policy.md

```diff
@@ -0,0 +1,329 @@
+
+The important change is that incremental updates must be inserted in to the map state *in priority order*. That is to say, higher-priority PerSelectorPolicies must apply their updates first. Then, the usual MapState deconfliction rules apply, and there is no risk of an incremental update diverging from policy intent.
+
+It is safe to insert incremental updates in this manner because identities and selectors are immutable. As a direct consequence, two selectors S1 and S2 might overlap, a new identity I1 can never change state from non-overlapping to overlapping.
+
+We will need to ensure that SelectorCache updates are processed in a transactional manner -- that is to say, an incremental update must be delivered to all affected selectors *before* distributing this to the MapState.
+
+#### Example 1: no wildcards
+
+Consider a set of rules:
+
+1. tcp 0-1023, ID (5,6)
+2. tcp 80, ID (7)
+3. tcp *, ID (8)
+
+Then an incremental update transaction adds ID 9. The sequence is thus:
+
+1. Rule 1 checks ID 9. ID 9 is not present in any ancestor rules. ID 9 is added.
+2. Rule 2 checks ID 9. ID 9 is present in an ancestor rule. ID 9 is **not added**.
+3. Rule 3 checks ID 9. There are no ancestor rules. ID 9 is added.
+
+
+```mermaid
+flowchart LR
+    subgraph Before
+    direction TB
+    rule1["prio 1, tcp, 0-1023, id {5, 6}"]
+    rule2["prio 2, tcp, 80, id {7}"]
+    rule3["prio 3, tcp, id {8}"]
+
+    root[[root]]
+    
+    root --> rule3
+    rule3 --> rule1
+    rule1 --> rule2
+    end
+    
+    subgraph After
+    direction TB
+    rule1a["prio 1, tcp, 0-1023, id {5, 6, 9}"]
+    rule2a["prio 2, tcp, 80, id {7, 9}"]
+    rule3a["prio 3, tcp, id {8}"]
+
+    roota[[root]]
+    
+    roota --> rule3a
+    rule3a --> rule1a
+    rule1a --> rule2a
+    end
+    
+    Before ~~~ After
+    
+    
+```
+
+#### Example 2: Wildcards
+
+
+## Impacts / Key Questions
+
+### KQ1: How should unordered and ordered policies interfact?
+
+To preserve existing behavior, all policies without an explicit order will have order number 0. Since policies with the same numerical order are merged, this preserves existing behavior.
+
+Only CCNPs are allowed to have a negative order number. This prevents namespace administrators from overriding cluster-wide policies.
+
+
+### Impact: effect on L7 policies
+
+For the time being, we will not permit L7 policies to specify an order. This is so that we do not need to implement an ordered L7-aware merge. It is likely this restriction will be lifted, once we gain familiarity with ordered policies.
+
+However, we still need to handle ordering between L4 and L7 policies.
+
+In the unordered case, any L7 policy overwrites a L4 policy. Consider two policies with the same selector:
+> - prio -1: allow access to `role=web, port 80` 
+> - prio 0: allow access to `role=web, port 80, host=example.com`
+
+In this case, the L7 policy would be ignored, since it is "overridden" by the L4 policy.
```

Comment:
What happens if the user supplies an explicit priority that inverts the core policy assumption of "any L7 policy overwrites an L4 policy"? In other words, what happens if the user does

```
> - prio 1: allow access to `role=web, port 80` 
> - prio 0: allow access to `role=web, port 80, host=example.com`
```

Reply:
That is the intention, yes. An L4 rule with higher priority would cause the L7 rule to be disregarded.

This CFP currently makes no statement as to the ordering of L7 rules. Logically, the expectation of

> - prio 0: allow access to host=example.com
> - prio 1: allow access to host=example.org

would be that example.com and example.org would be allowed. But that is not currently on the table. L7 policy must be priority 0.